subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/HistoricalTaskExecution.java

```diff
@@ -20,41 +20,49 @@
 import com.google.common.collect.ImmutableSortedMap;
 import com.google.common.collect.ImmutableSortedSet;
 import org.gradle.api.NonNullApi;
-import org.gradle.internal.id.UniqueId;
+import org.gradle.api.internal.tasks.OriginTaskExecutionMetadata;
 
 /**
  * State of a task when it was executed.
  */
 @NonNullApi
 public class HistoricalTaskExecution extends AbstractTaskExecution {
+
     private final boolean successful;
+    private final OriginTaskExecutionMetadata originExecutionMetadata;
     private final ImmutableSortedMap<String, FileCollectionSnapshot> inputFilesSnapshot;
     private final FileCollectionSnapshot discoveredInputFilesSnapshot;
     private final ImmutableSortedMap<String, FileCollectionSnapshot> outputFilesSnapshot;
 
     public HistoricalTaskExecution(
-        UniqueId buildInvocationId,
         ImplementationSnapshot taskImplementation,
         ImmutableList<ImplementationSnapshot> taskActionsImplementations,
         ImmutableSortedMap<String, ValueSnapshot> inputProperties,
         ImmutableSortedSet<String> outputPropertyNames,
         ImmutableSortedMap<String, FileCollectionSnapshot> inputFilesSnapshot,
         FileCollectionSnapshot discoveredInputFilesSnapshot,
         ImmutableSortedMap<String, FileCollectionSnapshot> outputFilesSnapshot,
-        boolean successful
+        boolean successful,
+        OriginTaskExecutionMetadata originExecutionMetadata
     ) {
-        super(buildInvocationId, taskImplementation, taskActionsImplementations, inputProperties, outputPropertyNames);
+        super(taskImplementation, taskActionsImplementations, inputProperties, outputPropertyNames);
         this.inputFilesSnapshot = inputFilesSnapshot;
         this.discoveredInputFilesSnapshot = discoveredInputFilesSnapshot;
         this.outputFilesSnapshot = outputFilesSnapshot;
         this.successful = successful;
+        this.originExecutionMetadata = originExecutionMetadata;
     }
 
     @Override
     public boolean isSuccessful() {
         return successful;
     }
 
+    @Override
+    public OriginTaskExecutionMetadata getOriginExecutionMetadata() {
```

Comment:
I think this method should be annotated with `@Nonnull`. Or is it possible to become null in `HistoricalTaskExecution`?

Reply:
Added